File: thesis/code/barter/gupea/src/barter/TradeAgent.java
package barter;

import java.io.IOException;
import java.io.InvalidObjectException;
import java.io.ObjectInputStream;
import java.util.List;

import sim.engine.SimState;
import sim.engine.Steppable;
import ec.util.MersenneTwisterFast;

/**
 * The agent class.
 * 
 * @NotThreadSafe This class is NOT thread safe!
 */
public final class TradeAgent implements Steppable,Comparable<TradeAgent> {

	private static final long serialVersionUID = 4247735360343951500L;

	private BarterStrategy barterStrategy;
	private ImprovementStrategy improvementStrategy;

	private int produceGood;
	private double score;
	private double produceAmount;

	private double[] price;			// price \equiv p_{this,goodIdx}
	private double[] demand;
	private double[] exchangeFor;
	private double[] inventory;
	private double[] consume;

	private transient TradeAgentProxy myProxy; // No need to serialize this

	/**
	 * Constructs an agent ready to start trading.
	 * 
	 * @param produceGood	the good that this agent produces
	 * @param price			the price vector for this agent
	 * @param params		the model parameters used to set up the agent
	 */
	public TradeAgent(int produceGood, double[] price, BarterParams params) {
		this.produceGood = produceGood;
		this.price = price.clone();
		int numGoods = params.getNumGoods();

		demand = new double[numGoods];
		exchangeFor = new double[numGoods];
		inventory = new double[numGoods];
		barterStrategy = new OriginalBarterStrategy();
		improvementStrategy = new CopyAndMutateImprovementStrategy();

		myProxy = new TradeAgentProxy(this);

		setConsume(params.getConsumeArray());
		produce();

		assert checkInvariants(); // invariants established?
	}

	private boolean checkInvariants() {
		if (consume.length != price.length
				|| consume.length != demand.length
				|| consume.length != exchangeFor.length
				|| consume.length != inventory.length
				|| produceGood < 0
				|| produceGood >= consume.length
				|| produceAmount != consume[produceGood] * consume.length
				|| score < 0
				|| myProxy == null
				|| barterStrategy == null
				|| improvementStrategy == null) {

			return false;
		}

		// TODO: can't check in Gintis' version
//		for (int g = 0; g < consume.length; g++) {
//			if (exchangeFor[g] != price[g] * demand[g] / price[produceGood]) {
//				return false;
//			}
//		}

		return true;
	}

	void produce() {
		// first reset the inventory for all goods
		for (int good = 0; good < price.length; good++) {
			inventory[good] = 0;
		}
		inventory[produceGood] = produceAmount;

		setDemandAndSupply();
		assert checkInvariants();
	}

	private void setDemandAndSupply() {
		double lambda = lambda();

		assert lambda == 1.0;

		for (int good = 0; good < price.length; good++) {
			if (good != produceGood) {
				demand[good] = consume[good] * lambda;
				exchangeFor[good] = price[good] * demand[good] / price[produceGood];
			}
		}

		demand[produceGood] = 0;
		exchangeFor[produceGood] = 0;
	}

	/**
	 * How much value agent wants to consume divided by how much value it has
	 * produced (according to its own prices).
	 */
	private double lambda() {
		double consumeValue = 0;

		// See section 3.2, eq. 2, incomeConstraint \equiv M^o
		double incomeConstraint = price[produceGood] * inventory[produceGood];

		// Compute denominator of eq. 4 from section 3.2.
		for (int good = 0; good < price.length; good++) {
			// See sec. 3.2, eq. 4
			// consumeValue += price[good] * consume[good];
			consumeValue += price[good] * inventory[good]; //REVERT-BUGS
		}

		// TODO: parameterize this (boolean allowExcessConsumption)
		if (incomeConstraint > consumeValue) {
			return 1;
		}

		return incomeConstraint / consumeValue; // See section 3.2, eq. 4, \equiv \lambda^*
	}

	/**
	 * Tells this agent to start trading.  
	 */
	@Override
	public void step(SimState state) {
		BarterEconomy be = (BarterEconomy) state;
		int maxTries = be.getParams().getMaxTries();
		double[] amounts = new double[2];

		for (int wantGood : be.getTradeOrder()) {
			if (demand[wantGood] == 0) {
				continue;
			}

			List<TradeAgent> producers = be.getProducers(wantGood);

			amounts[0] = demand[wantGood];
			amounts[1] = exchangeFor[wantGood];

			for (int i = 0; i < maxTries; i++) {
				TradeAgent responder = producers.get(be.random.nextInt(producers.size()));

				if (responder.acceptOffer(produceGood, amounts[1], amounts[0])) {

					// in and out amounts adjusted by initiator and responder

					// If resources of responder is depleted
					adjustAmounts(wantGood, 0, produceGood, 1, amounts);
					responder.adjustAmounts(produceGood, 1, wantGood, 0, amounts);

					Double before1 = null, before2 = null;
					assert (before1 = responder.getInventory(produceGood) + inventory[produceGood]) != null;
					assert (before2 = responder.getInventory(wantGood) + inventory[wantGood]) != null;

					exchangeGoods(wantGood, amounts[0], produceGood, amounts[1]);
					responder.exchangeGoods(produceGood, amounts[1], wantGood, amounts[0]);

					Double after1 = null, after2 = null;
					assert (after1 = responder.getInventory(produceGood) + inventory[produceGood]) != null;
					assert (after2 = responder.getInventory(wantGood) + inventory[wantGood]) != null;

					assert Math.abs(before1 - after1) < 0.02;
					assert Math.abs(before2 - after2) < 0.02;

					eat();
					responder.eat();

					// Are invariants preserved after a trade?
					assert checkInvariants();
					assert responder.checkInvariants();

					break;
				}
			}
		}
	}

	/**
	 * Exchanged amount is limited by two factors:
	 * - how much of the outgoing good do we have in inventory
	 * - how much of the incoming good do we actually want (demand)
	 * 
	 * @return Array with adjusted amounts (inAmount, outAmount)
	 */
	private void adjustAmounts(int inGood, int inIdx,
			int outGood, int outIdx, double[] amounts) {

		double outAmount = amounts[outIdx];
		double inAmount = amounts[inIdx];

		double haveAmount = inventory[outGood];
		if (outAmount > haveAmount) {
			// we are offering more than we have, so lets offer only the
			// amount that we have and request less by the same factor
			inAmount *= haveAmount / outAmount;
			outAmount = haveAmount;
		}

		// shouldn't be the case for offerer!
		double demandAmount = demand[inGood];
		if (inAmount > demandAmount) {
			// lets take only as much as we want and give less by the same
			// factor
			outAmount *= demandAmount / inAmount;
			inAmount = demandAmount;
		}

		//REVERT-BUGS is relevant if don't set demand and supply after getting next generation
		if (exchangeFor[inGood] < outAmount) {
			// we don't agree to give as much as requested, so lets give what
			// we can and take less by the same factor
			inAmount *= exchangeFor[inGood] / outAmount;
			outAmount = exchangeFor[inGood];
		}

		// to ensure that trade conditions have not been changed
		assert outAmount == 0.0 ||
			Math.abs(amounts[inIdx] / amounts[outIdx] - inAmount / outAmount) < 0.00001
					: amounts[inIdx] / amounts[outIdx] + " vs " + inAmount / outAmount;

		amounts[inIdx] = inAmount;
		amounts[outIdx] = outAmount;
	}

	/**
	 * Takes one good from inventory and adds another. Also sets new demand and
	 * supply for received good.
	 */
	private void exchangeGoods(int inGood, double inAmount, int outGood, double outAmount) {
		inventory[inGood] += inAmount;
		inventory[outGood] -= outAmount;

		demand[inGood] -= inAmount;

		// exchangeFor[inGood] = validateAmount(price[inGood] * demand[inGood] / price[outGood]);
		exchangeFor[inGood] -= outAmount; //REVERT-BUGS

		for (int good = 0; good < consume.length; good++) {
			inventory[good] = validateAmount(inventory[good]);
			demand[good] = validateAmount(demand[good]);
			exchangeFor[good] = validateAmount(exchangeFor[good]);
		}
	}

	private boolean acceptOffer(int offerGood, double offerAmount, double wantAmount) {
		return inventory[produceGood] > 0 && 
			barterStrategy.acceptOffer(myProxy,	offerGood, offerAmount, wantAmount);
	}

	private final void eat() {
		double min = Double.MAX_VALUE;

		// Calculate utility function (See section 3, eq. 1)
		for (int good = 0; good < price.length; good++) {
			min = Math.min(min, inventory[good] / consume[good]);
		}

		// If we have no inventory to consume, just return.
		if (min == 0.0) {
			return;
		}

		score += min;

		assert min <= 1 : "min > 1.0, was " + min;

		if (min < 1.0) {
			for (int good = 0; good < price.length; good++) {
				inventory[good] *= 1.0 - min;
			}
		} else {
			produce();
		}
	}

	private final static double validateAmount(double amount) {
		assert amount >= 0 : ("Negative amount: " + amount); 

		return amount < 0.01 ? 0.0 : amount;
	}

	// TODO: Could this be used for any malice/stupidity? Probably by the BarterStrategy! (Pelle)
	void resetScore() {
		score = 0.0;
	}

	/**
	 * Give this agent a chance to improve its prices based on a better agent. 
	 */
	public void improve(TradeAgent better,
			BarterParams params, MersenneTwisterFast random) {

		double[] newPrices =
			improvementStrategy.improve(better.getView(), myProxy, params, random);

		for (int good = 0; good < price.length; good++) {
			price[good] = newPrices[good];
		}

		assert checkInvariants();
	}

	/**
	 * Copy the strategy and also produced good of a better agent.
	 */
	void copy(TradeAgent better) {
		// TODO: Find a good way for the strategy to override this. (Pelle)

		for (int good = 0; good < price.length; good++) {
			price[good] = better.price[good];
			inventory[good] = better.inventory[good];
			demand[good] = better.demand[good];
			exchangeFor[good] = better.exchangeFor[good];
		}

		produceAmount = better.produceAmount;
		produceGood = better.produceGood;

		assert checkInvariants();
	}

	/**
	 * Returns the production good of this agent.
	 * @return The production good.
	 */
	public int getProduceGood() {
		return produceGood;
	}

	double getProduceAmount() {
		return produceAmount;
	}

	double getPrice(int good) {
		return price[good];
	}

	Class<? extends ImprovementStrategy> getImprovementStrategy() {
		return improvementStrategy.getClass();
	}
	
	Class<? extends BarterStrategy> getBarterStrategy() {
		return barterStrategy.getClass();
	}

	/**
	 * Returns the price array of this agent.
	 * @return The price array.
	 */
	public double[] getPrices() {
		// TODO: Can we do a defensive copy and yet have the ability
		// to graph the price from the MASON console? (Pelle)
		return price;
		//return price.clone();
	}

	double getDemand(int good) {
		return demand[good];
	}

	double getExchangeFor(int good) {
		return exchangeFor[good];
	}

	double getInventory(int good) {
		return inventory[good];
	}

	int getNumGoods() {
		return price.length;
	}

	/**
	 * Returns the score of this agent.
	 * @return The score.
	 */
	public double getScore() {
		return score;
	}

	/**
	 * Returns a limited view on this agent.
	 * @return The view on this agent.
	 */
	public TradeAgentProxy getView() {
		return myProxy;
	}

	void setConsume(double[] consume) {
		this.consume = consume.clone();
		produceAmount = consume[produceGood] * consume.length;
	}

	/**
	 * Sets the barter strategy for this agent.
	 * @param strategy	The barter strategy.
	 */
	public void setBarterStrategy(BarterStrategy strategy) {
		this.barterStrategy = strategy;
	}

	/**
	 * Sets the improvement strategy for this agent.
	 * @param strategy	The improvement strategy.
	 */
	public void setImprovementStrategy(ImprovementStrategy strategy) {
		this.improvementStrategy = strategy;
	}

	/**
	 * Compares this agent to another one. Returns 1 if this agent has better
	 * score, -1 if the score is lower, or 0 if they are equal.
	 */
	@Override
	public int compareTo(TradeAgent a) {
		return (int) -Math.signum(getScore() - a.getScore());			
	}

	private void readObject(ObjectInputStream s) throws IOException, ClassNotFoundException {
		s.defaultReadObject();

		myProxy = new TradeAgentProxy(this);

		// Check that our invariants are satisfied
		if (!checkInvariants()) {
			throw new InvalidObjectException("Invariants of TradeAgent not satisfied!");
		}
	}
}

File: thesis/code/barter/haskell/src/TradeAgent.hs
module TradeAgent
  ( TradeAgentState -- don't export internals of state

  , mkTradeAgent
  , getPrices
  , getScore
  , getProduceGood

  , step
  ) where

import Control.Monad.Random

import BarterParams
import Utils

-- TODO: need a generalisation of the barter-strategy
-- TODO: need a generalisation of the improvement-strategy

data TradeAgentState = TradeAgentState 
  { agentIndex    :: !Int

  , produceGood   :: !Int
  , score         :: !Double
  , produceAmount :: !Double

  , price         :: ![Double]
  , demand        :: ![Double]
  , exchangeFor   :: ![Double]
  , inventory     :: ![Double]
  , consume       :: ![Double]
  }

getPrices :: TradeAgentState -> [Double]
getPrices = price

getScore :: TradeAgentState -> Double
getScore = score

getProduceGood :: TradeAgentState -> Int
getProduceGood = produceGood

mkTradeAgent :: Int
             -> Int 
             -> [Double]
             -> BarterParams
             -> TradeAgentState
mkTradeAgent ai pg pr params = produce $ TradeAgentState
    { agentIndex    = ai

    , produceGood   = pg
    , score         = 0
    , produceAmount = prodAm

    , price         = pr
    , demand        = replicate n 0
    , exchangeFor   = replicate n 0
    , inventory     = replicate n 0
    , consume       = conArr
    }
  where
    n = getNumGoods params

    conArr = getConsumeArray params
    prodAm = conArr !! pg * fromIntegral (length conArr);

produce :: TradeAgentState -> TradeAgentState
produce s = setDemandAndSupply $ s { inventory = inventory'' }
  where
    n           = length $ price s
    inventory'  = replicate n 0
    inventory'' = replaceElem (produceGood s) (produceAmount s) inventory'
    -- TODO: check invariants

setDemandAndSupply :: TradeAgentState -> TradeAgentState
setDemandAndSupply s 
    = s { demand      = demand''
        , exchangeFor = exchangeFor'' }
  where
    l            = lambda s
    priceOfGood  = price s !! produceGood s

    demand'      = map (* l) (consume s) 
    exchangeFor' = zipWith (\p d -> p * d / priceOfGood) (price s) (demand s)

    demand''      = replaceElem (produceGood s) 0 demand'
    exchangeFor'' = replaceElem (produceGood s) 0 exchangeFor'

{-
  How much value agent wants to consume divided by how much value it has
  produced (according to its own prices).
-}
lambda :: TradeAgentState -> Double
lambda s 
    | incomeConstraint > consumeValue = 1
    | otherwise = incomeConstraint / consumeValue -- See section 3.2, eq. 4, \equiv \lambda^*
  where
    -- See section 3.2, eq. 2, incomeConstraint \equiv M^o
    incomeConstraint = price s !! produceGood s * inventory s !! produceGood s

    -- Compute denominator of eq. 4 from section 3.2.
    -- See sec. 3.2, eq. 4
    -- consumeValue = foldr (\(p, c) acc -> acc + p * c) 0 (zip (price s) (consume s))
    -- REVERT-BUGS
    consumeValue = foldr (\(p, c) acc -> acc + p * c) 0 (zip (price s) (inventory s))

-- start trading
step :: RandomGen g
     => BarterParams
     -> TradeAgentState 
     -> [TradeAgentState]
     -> Rand g (Maybe (TradeAgentState, TradeAgentState))
step params s0 ss = iterateGoodPermuation tradeGoodPermutation s0
  where
    maxRetries           = getMaxRetries params
    tradeGoodPermutation = getTradeOrder params

    iterateGoodPermuation :: RandomGen g
                          => [Int]
                          -> TradeAgentState
                          -> Rand g (Maybe (TradeAgentState, TradeAgentState))
    iterateGoodPermuation [] _ = return Nothing -- TODO return updated 
    iterateGoodPermuation (wantGood:gs) s
      | demand s !! wantGood == 0 = iterateGoodPermuation gs s
      | otherwise = findTrades maxRetries s wantGood producers
        where
          producers = filter (\s' -> produceGood s' == wantGood) ss 

    findTrades :: RandomGen g
               => Int
               -> TradeAgentState
               -> Int
               -> [TradeAgentState]
               -> Rand g (Maybe (TradeAgentState, TradeAgentState))
    findTrades 0 _ _ _ = return Nothing
    findTrades n s wantGood producers = do
      responder <- randomElem producers
      
      if not $ acceptOffer responder (produceGood s) (demand s !! wantGood) (exchangeFor s !! wantGood)
        then findTrades (n-1) s wantGood producers
        else return $ Just $ transactTrade wantGood s responder

    transactTrade :: Int 
                  -> TradeAgentState
                  -> TradeAgentState
                  -> (TradeAgentState, TradeAgentState)
    transactTrade wantGood s responder 
        = (s'', responder'')
      where
        amounts = [demand s !! wantGood, exchangeFor s !! wantGood]
        pg      = produceGood s

        (adjustedDemand, adustedExchange)   = adjustAmounts s wantGood 0 pg 1 amounts
        (adjustedDemand', adustedExchange') = adjustAmounts responder pg 1 wantGood 0 [adjustedDemand, adustedExchange]

        s'         = exchangeGoods wantGood adjustedDemand' pg adustedExchange' s
        responder' = exchangeGoods pg adustedExchange' wantGood adjustedDemand' responder

        s''         = eat s'
        responder'' = eat responder'

        -- TODO: check invariants

eat :: TradeAgentState -> TradeAgentState
eat s0
    | m == 0    = s0 -- If we have no inventory to consume, just return.
    | otherwise = s2
  where
    -- Calculate utility function (See section 3, eq. 1)
    m = foldr (\(i, c) acc -> min acc (i / c)) (1/0) (zip (inventory s0) (consume s0))

    -- assert min <= 1 : "min > 1.0, was " + min;

    s1 = s0 { score = score s0 + m }
    s2 = if m < 1
          then adjustInventory s1
          else produce s1

    adjustInventory s = s { inventory = inventory' }
      where
        inventory' = map (\i -> i * (1 - m)) (inventory s)

exchangeGoods :: Int
              -> Double
              -> Int 
              -> Double
              -> TradeAgentState
              -> TradeAgentState
exchangeGoods inGood inAmount outGood outAmount s0
    = s4 { inventory   = inv
         , demand      = dem
         , exchangeFor = exc }
  where
    s1 = changeInventory inGood inAmount s0
    s2 = changeInventory outGood (-outAmount) s1

    s3 = changeDemand inGood (-inAmount) s2
    
    -- s4 = _setExchange inGood (validateAmount $ price s3 !! inGood * demand s3 !! inGood / price s3 !! outGood) s3
    s4 = _changeExchange inGood (-outAmount) s3 -- REVERT-BUGS

    inv = map validateAmount (inventory s4)
    dem = map validateAmount (demand s4)
    exc = map validateAmount (exchangeFor s4)

    setStateArrayValue :: Int
                      -> Double
                      -> (TradeAgentState -> [Double])
                      -> ([Double] -> TradeAgentState)
                      -> TradeAgentState
                      -> TradeAgentState
    setStateArrayValue idx v proj updt s 
        = updt arr'
      where
        arr  = proj s
        arr' = replaceElem idx v arr

    _setExchange idx v s
      = setStateArrayValue idx v exchangeFor (\e -> s { exchangeFor = e }) s

    updateStateArrayValue :: Int
                          -> Double
                          -> (TradeAgentState -> [Double])
                          -> ([Double] -> TradeAgentState)
                          -> TradeAgentState
                          -> TradeAgentState
    updateStateArrayValue idx v proj updt s 
        = updt arr'
      where
        arr    = proj s
        arrVal = arr !! idx
        arr'   = replaceElem idx (arrVal + v) arr

    changeInventory idx v s 
        = updateStateArrayValue idx v inventory (\i -> s { inventory = i }) s
    changeDemand idx v s 
        = updateStateArrayValue idx v demand (\d -> s { demand = d }) s
    _changeExchange idx v s 
        = updateStateArrayValue idx v exchangeFor (\e -> s { exchangeFor = e }) s

adjustAmounts :: TradeAgentState
              -> Int
              -> Int
              -> Int
              -> Int
              -> [Double]
              -> (Double, Double)
adjustAmounts s inGood inIdx outGood outIdx amounts 
    = (outAmount3, inAmount3)
  where
    outAmount0 = amounts !! outIdx
    inAmount0  = amounts !! inIdx

    haveAmount = inventory s !! outGood
    (outAmount1, inAmount1) 
      = if outAmount0 > haveAmount
          -- we are offering more than we have, so lets offer only the
          -- amount that we have and request less by the same factor
          then (haveAmount, inAmount0 * (haveAmount / outAmount0)) -- inAmount *= haveAmount / outAmount; outAmount = haveAmount;
          else (outAmount0, inAmount0)

    -- shouldn't be the case for offerer!
    demandAmount = demand s !! inGood
    (outAmount2, inAmount2)
      = if inAmount1 > demandAmount
          -- lets take only as much as we want and give less by the same factor
          then (outAmount0 * (demandAmount / inAmount1), demandAmount) -- outAmount *= demandAmount / inAmount; inAmount = demandAmount;
          else (outAmount1, inAmount1)

    -- REVERT-BUGS is relevant if don't set demand and supply after getting next generation
    (outAmount3, inAmount3)
      = if exchangeFor s !! inGood < outAmount2
          -- we don't agree to give as much as requested, so lets give what
          -- we can and take less by the same factor
          then (exchangeFor s !! inGood, exchangeFor s !! inGood * inAmount2) -- inAmount *= exchangeFor[inGood] / outAmount; outAmount = exchangeFor[inGood];
          else (outAmount2, inAmount2)

          {-}
    // to ensure that trade conditions have not been changed
    assert outAmount == 0.0 ||
      Math.abs(amounts[inIdx] / amounts[outIdx] - inAmount / outAmount) < 0.00001
          : amounts[inIdx] / amounts[outIdx] + " vs " + inAmount / outAmount;

    amounts[inIdx] = inAmount;
    amounts[outIdx] = outAmount;
    -}

acceptOffer :: TradeAgentState 
            -> Int
            -> Double
            -> Double
            -> Bool 
acceptOffer s offerGood offerAmount wantAmount
    = inventory s !! pg > 0
    -- TODO: encapsulate as strategy: takes a TradeAgentState and returns Bool
    && not (demand s !! offerGood == 0 
           || exchangeFor s !! offerGood == 0
           || price s !! offerGood * offerAmount < price s !! pg * wantAmount)
  where
    pg = produceGood s

_checkInvariants :: TradeAgentState -> Bool
_checkInvariants s 
    =  n /= length (price s)
    || n /= length (demand s)
    || n /= length (exchangeFor s)
    || n /= length (inventory s)
    || pg < 0
    || pg >= n
    || produceAmount s /= consume s !! pg * fromIntegral n
    || score s < 0
  where
    n = length $ consume s
    pg = produceGood s